net: lib: Adding support for TFTP Client.

olsky opened the pull request:
Adding RFC1350 compliant support for TFTP Client in Zephyr. The
current implementation is minimal and only supports the ability
to get a file from the server.

Things for the future include support for putting files to
server and adding support for RFC2347.

tbursztyka commented on include/net/tftp.h:
```diff
@@ -0,0 +1,39 @@
+	u32_t  user_buf_size;
+};
+
+/* Name: tftp_get
```
as a public API, you'll need to properly document it via doxygen format (@brief, @param etc...)

bwasim replied:
Yes, I was wondering how the documentation part will be done.. Thanks for pointing this out.. 

BTW - Do you have any thoughts on the API, specifically the IO params ? Obviously the remote file has to a string but we could've kept the local file a string or an array. Though the former makes more sense (in the TFTP context) but it requires access to the File System+NVS and we wanted to keep our solution simple. Therefore we allowed the user to simply provide an array which can be populated by the stack.

pfalcon replied:
> BTW - Do you have any thoughts on the API, specifically the IO params ?

FYI, there's https://github.com/zephyrproject-rtos/zephyr/issues/13212 RFC for requirements on implementing higher-level networking protocols. In summary, it argues a) for always consider and support a "streaming" usecase (e.g., here transferring 1GB over TFTP and wrting it to an external flash); b) structure APIs calls as consistent as possible to standard Socket API, i.e. center around read(void *your_protocol, void *buf, size_t sz) and write(void *your_protocol, const void *buf, size_t sz) calls.

> but it requires access to the File System+NVS and we wanted to keep our solution simple.

Surely, there's no need for a networking protocol implementation to depend on a particular FS implementation. Users who need to work with an FS, can implement integration on their app side.

tbursztyka commented on include/net/tftp.h:
```diff
@@ -0,0 +1,39 @@
+ * Description: This function gets "file" from the remote server.
+ */
+int tftp_get(struct sockaddr_in *server, struct tftpc *client,
+		     const char *remote_file, const char *mode);
```
indentation is following linux one, so below last '(' + 1 space

jukkar replied:
indentation

tbursztyka commented on subsys/net/lib/tftp/Kconfig:
```diff
@@ -0,0 +1,64 @@
+	  request. This number dictates the number of times we will do re-tx of our request before giving up
+	  and exiting.
+
+endif # TFTP_LIB
```
you need to break a line after this (it usually breaks make menuconfig without it)

bwasim replied:
Added..

tbursztyka commented on subsys/net/lib/tftp/tftp_client.c:
I guess refactoring this whole loop content would be nicer, too many code block {} imbrication. We are supposed to follow linux code style, but I have seen patches being applied lately which were not following it (80 chars limit etc...) so what to trust or follow these days...

tbursztyka commented on subsys/net/lib/tftp/tftp_client.c:
same here

bwasim commented on subsys/net/lib/tftp/tftp_client.h:
```diff
@@ -0,0 +1,79 @@
+#define ERROR_OPCODE             0x5
+
+/* TFTP Client Success / Error Codes. */
+#define TFTPC_DATA_RX_SUCCESS     1
```
Anyone having thoughts on these error codes ? Some of these are directly conveyed to the application user and so they should be moved to the public `tftp.h` file I suppose ? 

In general, is there any guidelines to add new error codes (to ensure no overlap) or I can add whatever number I want ?

jukkar commented:
Hmm, I did not realized that you were creating a new PR so commented the old one (you could just have pushed your fixed tree on top of your older PR). Can you check my comments from the closed PR?

jukkar commented on include/net/tftp.h:
```diff
@@ -0,0 +1,39 @@
+/*
```
Please change the commit subject to `net: tftp: ...` in all of your commits.

olsky replied:
> Hmm, I did not realized that you were creating a new PR so commented the old one (you could just have pushed your fixed tree on top of your older PR). Can you check my comments from the closed PR?

----

Commenting this although you closed it. I would prefer that we have the IP address in the API in protocol agnostic way, so struct sockaddr *server here. Then it is easier to add IPv6 support too.

---
thanks!
yes, sorry something went wrong yesterday... we will fix it.

bwasim replied:
Done..

tbursztyka commented:
please fix the checkpatch issues.

rlubos commented on subsys/net/lib/tftp/tftp_client.c:
```diff
@@ -0,0 +1,524 @@
+	int     stat;
+
+	/* Enable select on the socket. */
+	pollfds_add(sock);
```
This should be done once, after the socket is created.

rlubos commented on subsys/net/lib/tftp/tftp_client.c:
```diff
@@ -0,0 +1,524 @@
+ * NUM_FDS be minus 2
+ */
+fd_set readfds;
+fd_set workfds;
```
Unused variable?

rlubos replied:
All remaining variables should be `static` - this helps to identify such cases easily.

bwasim replied:
Fixed..

rlubos commented on subsys/net/lib/tftp/tftp_client.c:
What about `readfds`, `fdnum`, we should initialize them as well.

bwasim replied:
All gone now as we're using `poll` instead of `select`..

rlubos commented on subsys/net/lib/tftp/tftp_client.c:
I'd re-think the logic here - a `goto` label within a loop is not an easy-to-follow solution.

bwasim replied:
I've refactored the logic to make it simpler.. Let me know what you think about it..

bwasim commented on subsys/net/lib/tftp/tftp_client.c:
```diff
@@ -0,0 +1,524 @@
+	.tv_usec  =   0
+};
+
+/* Number of simultaneous client connections will be
```
Yeah, I don't like this either and think this is something we can fix by using the "buffer" provided by the user in a better way and by creating all this "select" stuff on the stack.. Will try to experiment and commit if easily doable..

bwasim commented on subsys/net/lib/tftp/tftp_client.c:
TFTP is now "IP" version agnostic..

bwasim commented:
Can anyone tell me why Identity / Emails is failing ? I put in the Real Name / Email information to get rid of the problem but it still seems to be there..

bwasim committed:
net: tftp: Adding support for TFTP Client.

Adding RFC1350 compliant support for TFTP Client in Zephyr. The
current implementation is minimal and only supports the ability
to get a file from the server.

Things for the future include support for putting files to
server and adding support for RFC2347.

Signed-off-by: Bilal Wasim <bilalwasim676@gmail.com>

bwasim commented:
@jukkar / @tbursztyka / @pfalcon , please review !!

jukkar commented on subsys/net/lib/tftp/Kconfig:
```diff
@@ -0,0 +1,48 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config TFTP_LIB
+	bool "Socket TFTP Library Support"
```
Could you mark this as experimental, so just add `... [EXPERIMENTAL]` at the end of this text. This way we can more easily change the API when needed.

jukkar commented on subsys/net/lib/tftp/tftp_client.c:
```diff
@@ -0,0 +1,344 @@
+
+	/* Did we get some data? Or did we time out? */
+	if (stat > 0) {
+
```
Remove extra empty line

jukkar commented on subsys/net/lib/tftp/tftp_client.c:
```diff
@@ -0,0 +1,344 @@
+	block_no = get_u16(tftpc_buffer + 2);
+
+	if (tftpc_block_no > block_no) {
+
```
same here

jukkar commented on subsys/net/lib/tftp/tftp_client.c:
```diff
@@ -0,0 +1,344 @@
+
+	/* Only copy block if the user buffer has enough space. */
+	if (RECV_DATA_SIZE() > (client->user_buf_size - tftpc_index)) {
+
```
here too

jukkar commented on subsys/net/lib/tftp/tftp_client.c:
indentation

jukkar commented on subsys/net/lib/tftp/tftp_client.c:
indentation

bwasim replied:
What is the indentation supposed to be ? Should arguments align ? I looked at various NET files but wasn't able to find a conclusive answer. So I've aligned args..

jukkar replied:
The alignment for example for this function could look like this
```
static int tftp_send_request(int sock, u8_t request,
			     const char *remote_file, const char *mode)
```
so using a combination of tabs + spaces to align the arguments.
There are some exceptions to this if the argument is very long and would make the line >80 long.
Can you check also other functions in this file, I see multiple indentation issues. Other than that, looks good.

bwasim replied:
I have fixed all instances except `tftp_send_request` because the alignment stuff causes the line width to become greater than 80 and so checkpatch complains.. Even doing something like, 
`static int tftp_send_request(int sock, u8_t request,
			                     const char *remote_file,
                                             const char *mode)
`
also gives checkpatch errors.. So I've simply added 4 tabs here..

jukkar commented on subsys/net/lib/tftp/tftp_client.c:
remove extra empty line

jukkar commented on subsys/net/lib/tftp/tftp_client.c:
and here too

jukkar commented on subsys/net/lib/tftp/tftp_client.c:
remove extra empty line

jukkar commented on subsys/net/lib/tftp/tftp_client.c:
and here

jukkar commented on subsys/net/lib/tftp/tftp_client.c:
remove extra space at the end of the string

jukkar commented on subsys/net/lib/tftp/tftp_client.c:
```diff
@@ -0,0 +1,344 @@
+
+	if (tftpc_block_no > block_no) {
+
+		LOG_INF("Duplicate block received: %d", block_no);
```
I see lot of `LOG_INF()` messages printed in this code. Are you sure you want that, usually the INFO level is not very useful, and probably debug would be more appropriate.

bwasim replied:
Made all changes including this.. Let me know what more is required.. I will also be pushing a sample application once this PR is merged..

jukkar commented on subsys/net/lib/tftp/tftp_client.c:
indentation

jukkar commented on subsys/net/lib/tftp/tftp_client.c:
indentation

bwasim replied:
I don't fully understand why github is showing it this way because I have them properly aligned.. Couple of tabs and a single space to properly align `remote_file` under `server`..

pfalcon replied:
> I don't fully understand why github is showing it this way

Because it shows it in diff format, i.e. with "+" prepended. That may indeed cause mis-alignment on screen in the presence of tabs. But usually, such misalignment is small. Here, it looks like second line of the prototype has 3 tabs at the beginning. I would guess that's too much, there would be 1 tab + spaces to follow Zephyr style.

jukkar commented on subsys/net/lib/tftp/tftp_client.c:
```diff
@@ -0,0 +1,329 @@
+ * out directly to the TFTP server.
+ */
+static u32_t make_request(const char *remote_file, const char *mode,
+						  u8_t request_type)
```
indentation

jukkar commented on include/net/tftp.h:
```diff
@@ -0,0 +1,62 @@
+ * @return TBD
+ */
+int tftp_put(struct sockaddr *server, struct tftpc *client,
+		     const char *remote_file, const char *mode);
```
indentation

jukkar reviewed (commented):
Couple of indentation issues you missed.

olsky commented:
@jukkar thank you for your time and comments! We learn alot on the way. 
And moving from a pure proto to the next level :) takes some time and effort -- but we are coming to...

pfalcon commented on subsys/net/lib/tftp/tftp_client.c:
```diff
@@ -0,0 +1,329 @@
+	/* Enable select on this socket. Wait for the specified number
+	 * of seconds before exiting the call.
+	 */
+	stat = select(fdnum, &readfds, NULL, NULL, &tftpc_timeout);
```
I would strongly recommend against use of select(). It's a legacy function, which was added to Zephyr solely to make it easier to port 3rd-party apps to run with Zephyr. But IMHO, nothing in Zephyr tree should use it in their implementation (besides tests/samples). Instead, poll() should be used.

cc @jukkar, I don't think we aware discussed this explicitly, so don't voting it down right away, but IMHO, we should have such a rule.

jukkar replied:
> I don't think we aware discussed this explicitly, so don't voting it down right away, but IMHO, we should have such a rule.

Hmm, do not recall we have discussed this. But yes, the select() calls poll() so why not call poll() directly.

pfalcon replied:
> But yes, the select() calls poll() so why not call poll() directly.

Yes, in Zephyr, select() is implemented in terms of poll(), so using select() just adds more code and overhead for marshalling of fd_set <=> struct pollfd.

bwasim replied:
Thanks for the suggestion.. This has removed a lot of unnecessary / redundant code from the implementation.. Incorporated and tested..

pfalcon commented:
@olsky:

> thank you for your time and comments! We learn alot on the way.
And moving from a pure proto to the next level :) takes some time and effort -- but we are coming to...

Random question - why this PR was submitted by you (and still has your signoff in the description), but appears to be worked on by @bwasim ? (It's just unclear who's primary point of contact for this PR.)

olsky reviewed (commented):
> @olsky:
> 
> > thank you for your time and comments! We learn alot on the way.
> > And moving from a pure proto to the next level :) takes some time and effort -- but we are coming to...
> 
> Random question - why this PR was submitted by you (and still has your signoff in the description), but appears to be worked on by @bwasim ? (It's just unclear who's primary point of contact for this PR.)


We are committing with Bilal... why I am reviewer -- i dont know, definitely a mistake (hope not to have clicked on wrong button).

pfalcon commented:
> We are committing with Bilal...

Ack.

> why I am reviewer -- i dont know, definitely a mistake (hope not to have clicked on wrong button).

Well, I wouldn't worry about that, anybody can be a reviewer. (It's just some people, "codeowners", *have* to be reviewers.)

pfalcon commented:
The commit title:

> net: tftp: Adding support for TFTP Client. 

While I appreciate that commit title was already changes (perhaps a few times), my 2 cents is that it should start with "net: lib:" (because my first reaction to "net: tftp" was "we're adding tftp straight into the network stack??!!1").

olsky commented:
> The commit title:
> 
> > net: tftp: Adding support for TFTP Client.
> 
> While I appreciate that commit title was already changes (perhaps a few times), my 2 cents is that it should start with "net: lib:" (because my first reaction to "net: tftp" was "we're adding tftp straight into the network stack??!!1").

ov! okay, got the point, my mistake -- I just copied from other PRs thinking it will hint on the location and functionality.

thanks: btw!

olsky commented:
> > We are committing with Bilal...
> 
> Ack.
> 
> > why I am reviewer -- i dont know, definitely a mistake (hope not to have clicked on wrong button).
> 
> Well, I wouldn't worry about that, anybody can be a reviewer. (It's just some people, "codeowners", _have_ to be reviewers.)

okay, thanks! for me and us -- it is still a learning phase in zephyr and in such a huge project :) but we are on the way...

jukkar commented:
> my 2 cents is that it should start with "net: lib:" (because my first reaction to "net: tftp" was "we're adding tftp straight into the network stack??!!1").

I suggested the "net: tftp: .." to be used. I think it is just fine. The tokens in the commit subject are not members of a path but tell the component that is used. If the submitter wants to put :lib there, that is ok too but imho not really needed.

jukkar commented:
In the latest sent version, I still see the function indentation being wrong. I am hoping you are not editing the commits "inline" using github patch edit tool. I would recommend against it if you are doing it.

bwasim commented:
> In the latest sent version, I still see the function indentation being wrong. I am hoping you are not editing the commits "inline" using github patch edit tool. I would recommend against it if you are doing it.

@jukkar , This ( ![image](https://user-images.githubusercontent.com/19732152/80112084-2296fe00-859a-11ea-98fc-5c7d464b73b9.png) ) is my actual code which is perfectly aligned.. I don't understand why github is reporting otherwise.. Does github treat tabs differently ?

jukkar commented:
> This is my actual code which is perfectly aligned.. I don't understand why github is reporting otherwise.. Does github treat tabs differently ?

It is not github, I just fetched your repo and I see that all your functions are incorrectly aligned. I suspect your tab size is not 8, can you verify.
This is your file
```
static int tftp_send_request(int sock, u8_t request,
				const char *remote_file, const char *mode)
```
Which has 8 tabs, instead it should have 3 tabs and 5 spaces like this
```
static int tftp_send_request(int sock, u8_t request,
			     const char *remote_file, const char *mode)
```

bwasim commented:
> > This is my actual code which is perfectly aligned.. I don't understand why github is reporting otherwise.. Does github treat tabs differently ?
> 
> It is not github, I just fetched your repo and I see that all your functions are incorrectly aligned. I suspect your tab size is not 8, can you verify.
> This is your file
> 
> ```
> static int tftp_send_request(int sock, u8_t request,
> 				const char *remote_file, const char *mode)
> ```
> 
> Which has 4 tabs, instead it should have 3 tabs and 5 spaces like this
> 
> ```
> static int tftp_send_request(int sock, u8_t request,
> 			     const char *remote_file, const char *mode)
> ```

Thanks.. Fixed..

bwasim commented:
@jukkar / @tbursztyka / @pfalcon , please review !!

tbursztyka commented on subsys/net/lib/tftp/tftp_client.h:
```diff
@@ -0,0 +1,75 @@
+ * Description: This function gets u16_t buffer into a
+ * u8_t buffer array.
+ */
+static inline u16_t get_u16(u8_t *buf)
```
could you use instead sys_get_be16() from include/sys/byteorder.h

bwasim replied:
Great suggestion, Thanks.. I thought that such things would be available within net but never really looked..

tbursztyka commented on subsys/net/lib/tftp/tftp_client.h:
```diff
@@ -0,0 +1,75 @@
+ * Description: This function fills the provided u16_t
+ * buffer into a u8_t buffer array.
+ */
+static inline void put_u16(u8_t *buf, u16_t data)
```
and sys_put_be16() instead

tbursztyka commented on subsys/net/lib/tftp/tftp_client.c:
still indentation issue 
configure your editor to follow linux coding style

bwasim replied:
Done.. Won't it be preferable if Zephyr started using "spaces" instead of "tabs" because tabs can have different widths but spaces always have the same size ?

tbursztyka reviewed (changes requested):
change get_16/put_16 to sys_get_be16/sys_put_be16

jukkar commented on subsys/net/lib/tftp/tftp_client.c:
```diff
@@ -0,0 +1,306 @@
+	/* Enable poll on this socket. Wait for the specified number
+	 * of seconds before exiting the call.
+	 */
+	stat = poll(&fds, 1, K_SECONDS(CONFIG_TFTPC_REQUEST_TIMEOUT));
```
This will fail after we start to use k_timeout_t because K_SECONDS() will return k_timeout_t instead of int. In order to fix that just change the timeout to `MSEC_PER_SEC * CONFIG_TFTPC_REQUEST_TIMEOUT`

bwasim replied:
Fixed..

pfalcon replied:
> MSEC_PER_SEC * CONFIG_TFTPC_REQUEST_TIMEOUT

I'd suggest that having timeout in seconds is very coarse-grained, and would suggest to change it to be in milliseconds in the first place (figures like 0.5s and 1.5s are very valid for timeout).

bwasim replied:
Yeah, I have to agree with this.. Pushed and tested.. Now the Kconfig option is "msec" instead of "sec"..

pfalcon replied:
> Now the Kconfig option is "msec" instead of "sec"..

Thanks!

jukkar reviewed (commented):
Almost there, one issue with poll timeout

olsky commented:
@pfalcon @tbursztyka : can we finalize? are there anymore stoppers?

pfalcon commented on subsys/net/lib/tftp/tftp_client.c:
```diff
@@ -0,0 +1,306 @@
+#include "tftp_client.h"
+
+/* TFTP Global Buffer. */
+static u8_t   tftpc_buffer[TFTPC_MAX_BUF_SIZE];
```
I'd consider this bad design, this locks down non-trivial amount of memory (512+ bytes) for TFTP usage, which may never even happen. The way to solve this issue is to follow "dependency injection" pattern, where buffers are allocated and managed by the application, and `struct tftpc` contains a pointer to this buffer. This way, the same buffer can be used, in time-shared manner, for TFTP, HTTP, app's own processing, etc.

For reference, we have an RFC for guidelines on how to implement higher-level networking protocols: https://github.com/zephyrproject-rtos/zephyr/issues/13212 . This particular matter might not be there, but the point is that we do have some (proposed) guidelines.

Bottom line: I don't consider this to be worth -1, as TFTP is a special protocol, which would be used only for special needs (e.g. OTA-like). But I suggest to queue this up for a future refactor.

bwasim replied:
I had a couple of options when I started thinking about this problem, 

- **malloc buffer on user request** - When the user calls TFTP to get or put files from the server, we can dynamically allocate / deallocate buffers on every request. As this can cause memory fragmentation very quickly (if we have large number of tftp requests), this was not a very good solution. 
- **allocate buffers on the stack** - Instead of having a "global" internal buffer, we create a "512+4" bytes sized buffers on the stack but this imposes unnecessary burden on each calling task stack, I went against it as well.. 

Given all this, I decided to keep a global buffer which would get filled and eventually copied into the user buffer. I understand the problems that this approach has but it was simpler to implement and so I opted for this over the approach you've recommended here.. 

I had thought about implementing the approach mentioned by you a couple of weeks ago, but it requires "some" extra work and that's why I considered the current implementation to be okay for this version. Eventually we want to get rid of this "global" buffer as it is consuming large amount of memory and implement the mentioned approach..

pfalcon replied:
> I had thought about implementing the approach mentioned by you a couple of weeks ago, but it requires "some" extra work

Yes, and the proposed solution to this is to have "baseline" API which adheres to the best practices of e.g. resource usage, widest feature coverage, and composability, and on top of that, have an optional "convenience API", which has less knobs to control, and thus easier to use, at the expense of efficiency.

Example of that design is e.g. SNTP API: you can either call individual lower-level functions which allow you to have full control of what happens, or use a higher-level wrapper `sntp_simple()` https://github.com/zephyrproject-rtos/zephyr/blob/master/include/net/sntp.h#L98 which calls a number of functions for you. This way, we allow to develop with Zephyr both highly optimized and customized solutions, and at the same, code up typical usage scenarios quickly.

As I mentioned, I don't insist on this refactor for this PR (though it's pretty simple, and the drawback of not doing it now is breaking API changes later).

olsky replied:
your feedback is very valuable, in my opinion -- it is better to invest some time now, instead of taking that backpack with us and have a mountain to pass later. 
let us mitigate that part.

thanks!

jukkar replied:
> I had a couple of options when I started thinking about this problem,
> 
>     * **malloc buffer on user request** - When the user calls TFTP to get or put files from the server, we can dynamically allocate / deallocate buffers on every request. As this can cause memory fragmentation very quickly (if we have large number of tftp requests), this was not a very good solution.
> 
>     * **allocate buffers on the stack** - Instead of having a "global" internal buffer, we create a "512+4" bytes sized buffers on the stack but this imposes unnecessary burden on each calling task stack, I went against it as well..

There is a third option, just leave the allocations headache to the caller of this library. So the API would require that the caller supplies any extra buffers and lengths. This way we avoid extra management of buffers and stack usage, and application developer can optimize things in the application side. I think we discussed this issue earlier but you wanted to postpone this kind of change to later. Best of course would be to do that now as we have not yet released the API. The API is marked experimental so we can change it more easily later too.

Paul had a good point that we would need a sample app that just shows how the API can be used, so +1 to that.

pfalcon commented on subsys/net/lib/tftp/tftp_client.c:
```diff
@@ -0,0 +1,306 @@
+static u32_t  tftpc_buffer_size;
+
+/* TFTP Request block number property. */
+static u32_t  tftpc_block_no;
```
How presence of these global variables blends with:

1. Availability of `struct tftpc`.
2. Thus, apparent ability to have multiple `struct tftpc` in action.
3. Which can be executed concurrently in multiple threads.

?

bwasim replied:
All of these are correlated to the global "tftpc_buffer" and will go away once the tftp client implementation starts using the user supplied buffer in a better way..

pfalcon commented on subsys/net/lib/tftp/tftp_client.c:
Please remove this dummy function (returning success status) and copy-pasted incorrect comment.

pfalcon commented on subsys/net/lib/tftp/Kconfig:
```diff
@@ -0,0 +1,49 @@
+module-help=Enables tftp debug messages.
+source "subsys/net/Kconfig.template.log_config.net"
+
+config TFTPC_MAX_FILENAME_SIZE
```
I'm sorry, but this setting doesn't make sense. You do allocate a static buffer for entire TFTP packet. Until the suggestion to let dependency-inject this buffer from app side, you should have the config option which specifies the size of this static buffer. Then max filename size then just depends on this buffer size. And for current default for buffer size of 512, the max filename size is >500 bytes.

pfalcon reviewed (changes requested):
Sorry for bad news, but I see a few issues with the design of this TFTP client here. While some of them can be left for future refactor (as long as the client is marked as EXPERIMENTAL, so breaking changes are expected), I would like to request 2 changes right away:

1. Please remove dummy tftp_put() (because fixing it to be a proper non-implemented function is more effort than just removing it).
2. Please remove confusing CONFIG_TFTPC_MAX_FILENAME_SIZE option. Maximum TFTP filename size depends on the available TFTP packet buffer size, and nothing else.

pfalcon commented:
Another issue is that the PR lacks any test or sample, i.e. it's unclear whether it's known to actually work or not.

(I'm personally susceptible to promise from new contributors that "test/sample will come in next PR", but we all should understand: code which is not built (and that's what happens when there's no test or sample) means bitrotted code).

olsky commented:
> Another issue is that the PR lacks any test or sample, i.e. it's unclear whether it's known to actually work or not.
> 
> (I'm personally susceptible to promise from new contributors that "test/sample will come in next PR", but we all should understand: code which is not built (and that's what happens when there's no test or sample) means bitrotted code).

the tftp client was tested with nrf9160, @parthitce could you please share / push the sample FOTA ? as the sample was for NRF it's a wrong place here, but it might be we have a wifi test app too... @bwasim ?

pfalcon commented:
> the tftp client was tested with nrf9160, @parthitce could you please share / push the sample FOTA ? 

It's good to know. Not that ideal sample app would just test TFTP functionality and nothing (or as little as possible) else. So, it doesn't need (even shouldn't) be a FOTA example, rather FOTA can be plucked down to just part which downloads a file over TFTP (and ideally, checks that download is correct).  As an example, here's a sample which does all that for HTTP (via bare sockets even): https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/net/sockets/big_http_download/src/big_http_download.c

parthitce commented:
> So, it doesn't need (even shouldn't) be a FOTA example

@pfalcon I agree.

And the sample isn't clean and based on older revisions [here](https://github.com/parthitce/zephyr/commit/940ddb26d3d141780619848c3328945f1cb21d5e).

pfalcon commented:
> And the sample isn't clean and based on older revisions here.

Ok, so if you'd like to get this PR stabilized and merged first, and then refactor and submit a sample, I can certainly understand that, and it's ok by me.

pfalcon commented:
In the meantime, @bwasim, @olsky, any updates here? Two points I raised to get +1 from me aren't big, and release code freeze is around the corner (like, end of week), after which it may take extra effort to get get patch merge, and worst-case, it may be postponed until next merge window. (In the meantime, I assume you want to get this into the next release, so marked it as 2.3.0 milestone).

bwasim commented on subsys/net/lib/tftp/Kconfig:
```diff
@@ -0,0 +1,49 @@
+	  will assume that the Server failed and close the connection.
+
+config TFTPC_REQUEST_RETRANSMITS
+	int "Number of times the TFTP Client will retransmit the request and expect data from the server."
```
I'm made the description shorter..

bwasim commented on subsys/net/lib/tftp/Kconfig:
```diff
@@ -0,0 +1,49 @@
+	default 5
+	help
+	  Once the TFTP Client sends out a request, it will wait
+	  TFTPC_REQUEST_TIMEOUT secs for the data to arrive from the TFTP Server.
```
Yup, Fixed..

bwasim commented:
> In the meantime, @bwasim, @olsky, any updates here? Two points I raised to get +1 from me aren't big, and release code freeze is around the corner (like, end of week), after which it may take extra effort to get get patch merge, and worst-case, it may be postponed until next merge window. (In the meantime, I assume you want to get this into the next release, so marked it as 2.3.0 milestone).

Thanks.. I've removed the "TODO" tftp_put function and made the filename size bounded on the "tftpc_buffer".. 

Can you please review again ? Thanks !

pfalcon reviewed (approved):
Thanks for the updates, looks good!

pfalcon commented:
@tbursztyka: Can you please re-review from your side? (This still has your -1.)

bwasim commented:
The shippable failure is because of https://github.com/zephyrproject-rtos/zephyr/issues/14173 / https://github.com/zephyrproject-rtos/zephyr/issues/24635 and has nothing to do with TFTP..

pfalcon commented:
> The shippable failure is because 

Just close/reopen it to restart the build, to get around these random failures (yeah, they got too frequent recently :-( ).

bwasim commented:
@tbursztyka , Can you please review so we can merge ? Thanks..
